README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
jfsp
====

[![Travis-CI Build Status](https://travis-ci.org/leonawicz/jfsp.svg?branch=master)](https://travis-ci.org/leonawicz/jfsp) [![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/leonawicz/jfsp?branch=master&svg=true)](https://ci.appveyor.com/project/leonawicz/jfsp) [![Coverage Status](https://img.shields.io/codecov/c/github/leonawicz/jfsp/master.svg)](https://codecov.io/github/leonawicz/jfsp?branch=master)

The `jfsp` package encapsulates code, data and analysis results associated with ALFRESCO wildfire model outputs produced by the Scenarios Network for Alaska and Arctic Planning (SNAP) to assist the Joint Fire Science Program (JFSP). `jfsp` provides a collection of functions, data sets, apps and interactive documents for exploring and analyzing ALFRESCO wildfire model outputs associated with JFSP-related projects. It is a satellite package in the SNAPverse R package ecosystem.

Installation
------------

You can install jfsp from GitHub with:

``` r
# install.packages('devtools')
devtools::install_github("leonawicz/jfsp")
```


NEWS.md
# jfsp 0.6.6

* Updated `fbxfire` dataset, fixing erroneous values due to a bug in upstream data processing.

# jfsp 0.6.5

* Added a more robust and complete unit test suite for plots, dataset availability and dataset report generation.

# jfsp 0.6.1

* Added optional argument `obs = TRUE` to `jfsp_plot` for adding horizontal line overlay for mean historical observed values pertaining to BA SD and management cost plots.

# jfsp 0.6.0

* Added statewide Alaska burn area through time conditional on coniferous vs. deciduous tree species on the landscape.
* Updated `jfsp_plot` to offer a `cdba` plot option.
* Updated documentation.

# jfsp 0.5.0

* Updated all data sets. Added `labels` and `shortDescription` attributes to variables in data frames for use in external data set documentation.
* Revised data prep script with function to assist in variable attribution.

# jfsp 0.4.0

* Updated several data sets.

# jfsp 0.3.5

* Updated `jfsp_plot` to offer optional argument to triplicate annual historical data for inter-annual variability in burn area and prepend to each projected RCP series.
* Added `pt_size` argument for controlling point size in applicable plot types.
* Made optional argument `n` for moving average plot applicable to all versions of plot. Now defaults to `n = 30`.
* Updated data sets, making column names more consistent and informative.

# jfsp 0.3.0

* Added decadal `firesize` data set.
* Added fire size distributions plot to `jfsp_plot` type options.
* Refactored `jfsp_plot` to handle RCP marginalization and Treatment dropping via `by_rcp` and `by_tx`.
* Additional minor updates.

# jfsp 0.2.0

* Refactored `jfsp_plot`, making data sets implicit so that specifying them via argument is not necessary. Reduced, reordered formal arguments.
* Added optional arguments, including statewide aggregation of `fmoba` data for applicable plots using `alaska = TRUE`.

# jfsp 0.1.0

* Added primary plotting function for package data sets.
* Added seven package data sets.
* Update `jfsp_plot` function. It now handles plotting in R and saving plots to disk with an optional `file` argument. Some minimal handling for `showtext` issues when saving high dpi plots was also added.
* Updated unit tests.
* Updated documentation.

# jfsp 0.0.0.9000

* Initialized package and added packaged scaffolding.


tests/testthat.R
library(testthat)
library(jfsp)

test_check("jfsp")


tests/testthat/test-data.R
context("data")

x <- list(basecost, cost, costSummary, fmoba, fmobaSummary, cdratio, cdba, fbxfire, firesize)

test_that("data sets are available", {
  purrr::walk(x, ~expect_is(.x, "tbl_df"))
})


tests/testthat/test-docs.R
context("docs")

x <- list(basecost, cdratio, fbxfire, fmoba, fmobaSummary, cost, costSummary, firesize)

test_that("datadoc runs without error", {
  skip_on_appveyor()
  expect_is(datadoc(basecost, open_doc = FALSE), "NULL")
  expect_is(datadoc(cdratio, "out.html", open_doc = FALSE), "NULL")
  expect_is(datadoc(fbxfire, "out.pdf", open_doc = FALSE), "NULL")
  expect_is(datadoc(fmoba, "out.docx", open_doc = FALSE), "NULL")
  expect_is(datadoc(fmobaSummary, title = "title", open_doc = FALSE), "NULL")
  expect_is(datadoc(cost, subtitle = "subtitle", open_doc = FALSE), "NULL")
  expect_is(datadoc(costSummary, metadata = TRUE, open_doc = FALSE), "NULL")
  expect_is(datadoc(firesize, keep_rmd = TRUE, open_doc = FALSE), "NULL")
  expect_is(datadoc(cdba, details = FALSE), "NULL")
  unlink(c("firesize.Rmd", list.files(pattern = ".html"), "out.pdf", "out.docx"))
})


data-raw/data.R
library(dplyr)

f <- function(x) switch(x, historical = "Historical",
                        rcp45 = "RCP 4.5", rcp60 = "RCP 6.0", rcp85 = "RCP 8.5")
lev <- c("Historical", "RCP 4.5", "RCP 6.0", "RCP 8.5")

basecost <- tibble::data_frame(
  limited = c(13.9, 0.3, 7.6, 92.0, 10.6, 8.1, 12.0),
  modified = c(16, 22, 164, 356, 31, 56, 534),
  critical = c(2459, 473, 12868, 11160, 198, 6935, 8439),
  full = c(351, 257, 456, 76, 89, 312, 120)
)

# Alaska fire size distributions
firesize <- readRDS("C:/github/jfsp-archive/data-raw/fs.rds") %>% select(-Set) %>%
  mutate(Decade = Year - Year %% 10) %>% group_by(Tx, RCP, Decade, Val, Model) %>%
  summarise(Freq = sum(Freq)) %>% summarise(Freq = mean(Freq)) %>% ungroup %>%
  mutate(Tx = ifelse(Tx == "tx0", "Status quo", ifelse(Tx == "tx1", "Treatment 1", "Treatment 2")),
         Val = as.integer(round(247.105 * Val))) %>% rename(FS = Val)

# Alaska confierous:deciduous ratios
cdratio <- readRDS("C:/github/jfsp-archive/data-raw/conif_decid_area.rds") %>% select(-FMO) %>%
  group_by(Tx, RCP, Year, Vegetation) %>% summarise(Val = mean(Val)) %>%
  summarise(Val = Val[1] / Val[2]) %>% ungroup %>% rename(value = Val) %>%
  mutate(Tx = ifelse(Tx == "tx0", "Status quo", ifelse(Tx == "tx1", "Treatment 1", "Treatment 2")))

# Alaska confierous:deciduous burn area
cdba <- readRDS("C:/github/jfsp-archive/data-raw/conif_decid_ba.rds") %>% select(-FMO) %>%
  group_by(Tx, RCP, Year, Vegetation) %>% summarise(Val = mean(Val)) %>%
  ungroup %>% rename(value = Val) %>%
  mutate(value = as.integer(round(247.105 * value)),
         Tx = ifelse(Tx == "tx0", "Status quo", ifelse(Tx == "tx1", "Treatment 1", "Treatment 2")))

# Robust annual P(Fire) near Fairbanks, Alaska
fbxfire <- readRDS("C:/github/jfsp-archive/data-raw/fire_prob_fbks_simplified.rds") %>%
  mutate(RCP = factor(sapply(as.character(RCP), f), levels = lev))

gcms <- c("CCSM4", "GFDL-CM3", "GISS-E2-R", "IPSL-CM5A-LR", "MRI-CGCM3")
fmoba <- readRDS("C:/github/jfsp-archive/data-raw/ba_fmo/ba_fmo2_sw_alaska.rds")
sets <- strsplit(fmoba$Set, "\\.")
fmoba <- mutate(fmoba, Set = sapply(sets, "[", 1), RCP = sapply(sets, "[", 2),
                Model = gsub("CRU32", "CRU 3.2", sapply(sets, "[", 3))) %>%
  filter(Tx != "none") %>% mutate(RCP = factor(sapply(as.character(RCP), f), levels = lev)) %>%
  mutate(Tx = ifelse(is.na(Tx) & Set == "fmo99s95i", "Status quo",
                     ifelse(is.na(Tx), "No management",
                            ifelse(Tx == "tx0", "Status quo",
                                   ifelse(Tx == "tx1", "Treatment 1",
                                          ifelse(Tx == "tx2", "Treatment 2",
                                                 ifelse(Tx == "none", "No management", Tx)))))))
fmoba2 <- mutate(readRDS("C:/github/jfsp-archive/data-raw/ba_fmo/ba_fmo_sw_alaska.rds"),
             Set = substr(Set, 1, 9), RCP = factor("Historical", levels = lev),
             Model = "Observed", Tx = "Status quo")
fmoba <- bind_rows(fmoba, fmoba2) %>%
  mutate(Set = factor(Set, levels = c("Observed", "fmo99s95i")),
         Model = factor(Model, levels = c("Observed", "CRU 3.2", gcms)),
         Tx = factor(Tx, levels = c("No management", "Status quo", "Treatment 1", "Treatment 2"))) %>%
  select(c(1, 2, 10:11, 3:9)) %>% mutate_at(6:11, ~round(247.105 * .x)) %>%
  arrange(Set, Tx, RCP, Model, Year)

mgmtcost <- filter(fmoba, Set == "fmo99s95i")
fmoba <- tidyr::gather(fmoba, "FMO", "BA", -c(1:5)) %>%
  mutate(FMO = factor(FMO, levels = c("Unmanaged", "Limited", "Modified", "Full", "Critical", "Other")))
fmoba <- group_by(fmoba, Set, RCP, Model, Tx, FMO) %>%
  arrange(Set, Tx, RCP, Model, FMO, Year) %>% mutate(CBA = cumsum(BA)) %>% ungroup

# Summary statistics for FMO zone-specific burn area over time
fmobaSummary <- group_by(fmoba, Set, Tx, RCP, FMO) %>%
  summarise(min = min(BA), q1 = quantile(BA, 0.25), median = median(BA),
            mean = mean(BA), q3 = quantile(BA, 0.75), max = max(BA)) %>% ungroup

# FMO zone-specific annual burn area, plus cumulative and 10-year MA SD
fmoba <- group_by(fmoba, Set, Tx, RCP, FMO, Year) %>%
  summarise(BA = mean(BA), CBA = mean(CBA)) %>%
  mutate(BA_sd_ma10 = RcppRoll::roll_sd(BA, 10, fill = NA)) %>% ungroup

set.seed(1)
x <- replicate(1000,
  rowSums(mgmtcost[, 6:11]*cbind(0, sapply(basecost, sample, nrow(mgmtcost), replace = TRUE), 0))) / 1e6
cost <- mutate(mgmtcost, `5th percentile` = apply(x, 1, quantile, prob = 0.05),
               Mean = rowMeans(x), `95th percentile` = apply(x, 1, quantile, prob = 0.95)) %>%
  tidyr::gather("cost", "value", 12:14, factor_key = TRUE) %>%
  group_by(Tx, RCP, Year, cost) %>% summarise(value = mean(value))
costSummary <- mutate(mgmtcost, Decade = Year - Year %% 10) %>%
  bind_cols(as_data_frame(x)) %>% filter(Year >= 2020) %>% tidyr::gather("cost", "value", 13:1012) %>%
  group_by(Set, Tx, RCP, Decade) %>%
  summarise(`5th percentile` = quantile(value, prob = 0.05),
            Mean = mean(value), `95th percentile` = quantile(value, prob = 0.95)) %>% ungroup %>%
  select(-Set)

# set variable labels and attributes
set_atts <- function(data, lab, desc){
  x <- names(data)
  for(i in 1:ncol(data)){
    if(!is.na(lab[i])) attr(data[[x[i]]], "labels") <- lab[i]
    if(!is.na(desc[i])) attr(data[[x[i]]], "shortDescription") <- desc[i]
  }
  data
}

set <- c("Observed/ALFRESCO run", "Historical observations or ALFRESCO output based on various run parameterizations")
tx <- c("Control/Treatment run", "The control group is 'status quo' fire management and other values are different fire management treatments")
rcp <- c("Representative concentration pathways", "RCPs based on different greenhouse gas emissions scenarios")
fmo <- c("Fire management options", "Fire management options refer to regions of Alaska where a particular management strategy is applied")
yr <- c("Year", "")
dec <- c("Decade", "")
ba <- c("Burn area", "The annual area burned")
cba <- c("Cumulative burn area", "The cumulative annual area burned")
basd <- c("10-yr MA burn area SD", "The inter-annual variability in area burned given by the 10-year moving average of the standard deviation in burn area")

lab <- c(set[1], tx[1], rcp[1], fmo[1], yr[1], ba[1], cba[1], basd[1])
desc <- c(set[2], tx[2], rcp[2], fmo[2], yr[2], ba[2], cba[2], basd[2])
fmoba <- set_atts(fmoba, lab, desc)

lab <- c(set[1], tx[1], rcp[1], fmo[1], "Minimum", "First quartile", "Median", "Mean", "Third quartile", "Maximum")
desc <- c(set[2], tx[2], rcp[2], fmo[2], rep("Summary statistic: all years", 6))
fmobaSummary <- set_atts(fmobaSummary, lab, desc)

lab <- c(tx[1], rcp[1], yr[1], "Management cost statistic", "Management cost")
desc <- c(tx[2], rcp[2], yr[2], "5th percentile, mean, and 95th percentile statistics for the distribution of MC cost samples", "Fire management cost in millions of dollars")
cost <- set_atts(cost, lab, desc)

lab <- c(tx[1], rcp[1], dec[1], "Lower bound cost", "Mean cost", "Upper bound cost")
desc <- c(tx[2], rcp[2], dec[2], "5th percentile fire management cost in millions of dollars", "Mean fire management cost in millions of dollars", "95th percentile fire management cost in millions of dollars")
costSummary <- set_atts(costSummary, lab, desc)

lab <- c(tx[1], rcp[1], "Time period", "Spatial buffer", "P(Fire)")
desc <- c(tx[2], rcp[2], "Aggregate time period in years", "Spatial buffer radius in km", "The probability of fire near Fairbanks, Alaska in ALFRESCO output")
fbxfire <- set_atts(fbxfire, lab, desc)

lab <- c(tx[1], rcp[1], yr[1], "Coniferous:Deciduous ratio")
desc <- c(tx[2], rcp[2], yr[2], "The ratio of coniferous to deciduous species on the Alaska landscape through time in ALFRESCO output")
cdratio <- set_atts(cdratio, lab, desc)

lab <- c(tx[1], rcp[1], yr[1], "Coniferous:Deciduous burn area")
desc <- c(tx[2], rcp[2], yr[2], "TConiferous and deciduous species burn area on the Alaska landscape through time in ALFRESCO output")
cdba <- set_atts(cdba, lab, desc)

lab <- c(tx[1], rcp[1], dec[1], "Fire size", "Frequency")
desc <- c(tx[2], rcp[2], dec[2], "Fir size distributions per year in acres with a minimum of resolution of 247 acres after conversion to acres from native 1-km resolution ALFRESCO output", "Frequency of fires of a given size per year in ALFRESCO output")
firesize <- set_atts(firesize, lab, desc)

lab <- c("Limited management", "Modified management", "Critical management", "Full management")
desc <- c("Historical annual empirical fire management costs per unit area (dollars/acre) for the years 2001 - 2017")
basecost <- set_atts(select(basecost, c(1, 2, 4, 3)), lab, desc)

usethis::use_data(basecost, cdratio, cdba, fbxfire, fmoba, fmobaSummary, cost, costSummary, firesize)


R/datadoc.R
#' Generate a report on a data set
#'
#' Generate a pdf, Word or HTML report on any data set in the jfsp package.
#'
#' Create a report for a specific data set. The report summarizes the variables in the data set. It is a wrapper around and based on \code{dataMaid::makeDataReport}, though it renders output with some differences not available in \code{dataMaid}.
#'
#' Pandoc must be available on the system for rendering to pdf. It is recommended to run from RStudio, which includes the necessary components, if you have difficulty.
#' For Word documents (docx) it is necessary to be on a Windows system with Microsoft Word installed.
#' If no file name is supplied, the output defaults to HTML and the file name is the name of the data set.
#'
#' If \code{metadata = TRUE}, information about the system environment and conditions at run time when the document was generated is included at the bottom of the document. If \code{FALSE} (default), they are removed.
#' Because \code{datadoc} does not directly render the document with \code{dataMaid} and makes some internal changes to the Rmd file, the function call line in the metadata is always removed due to its resulting lack of clarity.
#'
#' @param data jfsp data set, e.g., \code{fmoba}.
#' @param file character, output file, must end in .pdf, .docx, or .html. See details.
#' @param title character, document title.
#' @param subtitle character, document subtitle. Set to \code{NULL} to recover the \code{dataMaid} attribution.
#' @param metadata logical, if \code{TRUE}, include metadata. See details.
#' @param open_doc logical, open document after rendering. Defaults to \code{TRUE}.
#' @param keep_rmd logical, retain the intermediary Rmd document. Defaults to \code{FALSE}.
#' @param details logical, if set to \code{FALSE}, do not print any progress.
#'
#' @return nothing is returned, but a file is written to disk.
#' @export
#'
#' @examples
#' \dontrun{datadoc(fmoba)}
datadoc <- function(data, file, title = "", subtitle = "", metadata = FALSE,
                    open_doc = TRUE, keep_rmd = FALSE, details = TRUE){
  data_name <- deparse(substitute(data))
  if(missing(file)) file <- paste0(data_name, ".html")
  ext <- utils::tail(strsplit(file, "\\.")[[1]], 1)
  if(!ext %in% c("pdf", "docx", "html")) stop("`file` must end in .pdf, .docx or .html.")
  rmd <- gsub(ext, "Rmd", file)
  if(title == "") title <- data_name
  if(details) cat("Genrating Rmd for ", data_name, "...\n", sep = "")
  render <- if(ext == "docx") TRUE else FALSE
  suppressMessages({
    dataMaid::makeDataReport(data, gsub("docx", "word", ext), render = render,
                             file = rmd, replace = TRUE, openResult = FALSE,
                             codebook = TRUE, reportTitle = paste0("\"", title, "\""))
  })
  if(details) cat("Rendering ", ext, " report...\n", sep = "")
  .intercept_report(rmd, subtitle, metadata)
  if(!keep_rmd) unlink(rmd)
  if(open_doc) pander::openFileInOS(file)
  if(details) cat("Completed.\n")
  invisible()
}

.intercept_report <- function(file, subtitle = "", metadata = FALSE){
  file_lines <- readLines(file)
  file_lines <- gsub("# Codebook summary table", "# Variable summary", file_lines)
  idx_stl <- which(substr(file_lines, 1, 9) == "subtitle:")
  idx_thm <- which(substr(file_lines, 1, 16) == "library(ggplot2)")
  idx_end <- which(substr(file_lines, 1, 30) == "Report generation information:")
  idx_end2 <- which(substr(file_lines, 1, 18) == " *  Function call:")
  file_lines[idx_thm] <- paste0("library(ggplot2); theme_set(theme_bw())")
  if(length(idx_stl) & !is.null(subtitle)) file_lines[idx_stl] <- paste0("subtitle: \"", subtitle, "\"")
  n <- length(file_lines)
  file_lines <- if(metadata) file_lines[-c(idx_end2:n)] else file_lines[-c(idx_end:n)]
  writeLines(file_lines, file)
  rmarkdown::render(file, quiet = TRUE)
  invisible()
}
